pkg/apis/core/v1beta1/helper/helper.go

```diff
@@ -782,3 +699,126 @@ func FindPrimaryDNSProvider(providers []gardencorev1beta1.DNSProvider) *gardenco
+// In case it does not find a version, it returns false.
+func GetKubernetesVersionForMinorUpdate(cloudProfile *gardencorev1beta1.CloudProfile, currentVersion string) (bool, string, error) {
+	currentSemVerVersion, err := semver.NewVersion(currentVersion)
+	if err != nil {
+		return false, "", err
+	}
+
+	latestVersion, err := GetLatestQualifyingVersion(cloudProfile.Spec.Kubernetes.Versions, FilterForConsecutiveMinor(*currentSemVerVersion), FilterSameVersion(*currentSemVerVersion), FilterExpiredVersionPredicate())
+	if err != nil {
+		// in case there are only expired versions in the consecutive minor version, pick the latest expired version
+		latestVersion, err = GetLatestQualifyingVersion(cloudProfile.Spec.Kubernetes.Versions, FilterForConsecutiveMinor(*currentSemVerVersion), FilterSameVersion(*currentSemVerVersion))
+		if err != nil {
+			return false, "", nil
+		}
+	}
+
+	return true, latestVersion.Version, nil
+}
+
+// GetLatestQualifyingVersion returns the latest MachineImage version from a set of expirable versions
+// A version qualifies if it's classification is not preview and the optional predicate does not filter out the version.
+// If the predicate returns true, the version is not considered for the latest qualifying version.
+func GetLatestQualifyingVersion(versions []gardencorev1beta1.ExpirableVersion, predicate ...func(expirableVersion gardencorev1beta1.ExpirableVersion, version *semver.Version) (bool, error)) (gardencorev1beta1.ExpirableVersion, error) {
+	filteredVersions := make(map[string]gardencorev1beta1.ExpirableVersion, len(versions))
+OUTER:
+	for _, v := range versions {
+		if v.Classification != nil && *v.Classification == gardencorev1beta1.ClassificationPreview {
+			continue
+		}
+
+		for _, p := range predicate {
+			if p == nil {
+				continue
+			}
+			semver, err := semver.NewVersion(v.Version)
+			if err != nil {
+				return gardencorev1beta1.ExpirableVersion{}, fmt.Errorf("error while parsing version '%s': %s", v.Version, err.Error())
+			}
+			shouldFilter, err := p(v, semver)
+			if err != nil {
+				return gardencorev1beta1.ExpirableVersion{}, fmt.Errorf("error while evaluation predicate: '%s'", err.Error())
+			}
+			if shouldFilter {
+				continue OUTER
+			}
+		}
+
+		filteredVersions[v.Version] = v
+	}
+	if len(filteredVersions) == 0 {
+		return gardencorev1beta1.ExpirableVersion{}, fmt.Errorf("no latest qualifying version could be determined")
+	}
+	latestVersion := sets.StringKeySet(filteredVersions).List()[len(filteredVersions)-1]
+	return filteredVersions[latestVersion], nil
+}
+
+// FilterForConsecutiveMinor returns a closure that evaluates whether a given version v has a consecutive minor version compared to the currentSemVerVersion
+// returns true if v has a consecutive minor version
+func FilterForSameMajorMinorVersion(currentSemVerVersion semver.Version) func(expirableVersion gardencorev1beta1.ExpirableVersion, version *semver.Version) (bool, error) {
```

Comment:
WDYT about introducing a type alias for the predicates? You write the func type out many times now.

```suggestion
func FilterForSameMajorMinorVersion(currentSemVerVersion semver.Version) VersionPredicate {
```